src/libopenrave-core/jsonparser/jsonreader.cpp

```diff
@@ -708,47 +708,47 @@ class JSONReader
             }
         }
 
+        KinBody::KinBodyInfoPtr pNewKinBodyInfo;
         if( insertIndex >= 0 ) {
             RAVELOG_DEBUG_FORMAT("env=%d, loaded referenced body '%s' with id='%s' from uri '%s'. Scope is '%s'", _penv->GetId()%envInfo._vBodyInfos.at(insertIndex)->_name%originBodyId%referenceUri%currentFilename);
-        }
-        else {
-            KinBody::KinBodyInfoPtr pNewKinBodyInfo;
+            pNewKinBodyInfo = envInfo._vBodyInfos[insertIndex];
+        } else {
             bool isNewRobot = orjson::GetJsonValueByKey<bool>(rRefKinBodyInfo, "isRobot", false);
             if( isNewRobot ) {
                 pNewKinBodyInfo.reset(new RobotBase::RobotBaseInfo());
             }
             else {
                 pNewKinBodyInfo.reset(new KinBody::KinBodyInfo());
             }
-            pNewKinBodyInfo->DeserializeJSON(rRefKinBodyInfo, fRefUnitScale, _deserializeOptions);
+        }
+        pNewKinBodyInfo->DeserializeJSON(rRefKinBodyInfo, fRefUnitScale, _deserializeOptions);
```

Comment:
there are code paths above that do:
```
envInfo._vBodyInfos.at(insertIndex)->DeserializeJSON(rRefKinBodyInfo, fUnitScale, _deserializeOptions);
```

and 
```
pExistingBodyInfo->DeserializeJSON(rRefKinBodyInfo, fRefUnitScale, _deserializeOptions);
```

So why aren't those code paths being hit? Potentially your change can lead to double deserialization.

Reply:
It's because `insertIndex` was kept at `-1` when the uri is local `#something`, since we didn't insert anything into `envInfo._vBodyInfos`

Because of `insertIndex = -1`, even though we found the `rRefKinBodyInfo`, we didn't `DeserializeJSON` with it.

Reply:
```
envInfo._vBodyInfos.at(insertIndex)->DeserializeJSON(rRefKinBodyInfo, fUnitScale, _deserializeOptions);
```
will not be hit because it's under this branch `else if (!scheme.empty() && !path.empty()) {...}`, and the uri just has `#somebodyID`

Reply:
```
pExistingBodyInfo->DeserializeJSON(rRefKinBodyInfo, fRefUnitScale, _deserializeOptions);
```
will not be hit because `insertIndex` is 0 at this time.

Reply:
@ziyan my issue is when `insertIndex >= 0` and hits that code path, then there is a possibility of it deserializing twice.  When `insertIndex=-1`, there's no change in behavior.
So why when `insertIndex=0` did we not hit the above code paths that deseriaize?

Reply:
@rdiankov @LLipter see my latest commit

Reply:
Pipeline #571680 testing